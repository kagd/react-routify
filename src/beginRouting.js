import page from 'page'

const beginRouting = () => {
  page()
}

export default beginRouting
