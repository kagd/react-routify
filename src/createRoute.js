import page from 'page'

const createRoute = (path, handler) => {
  page(path, handler)
}

export default createRoute
