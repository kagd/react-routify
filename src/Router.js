import { Component, PropTypes, createElement } from 'react'
import createRoute from './createRoute'
import beginRouting from './beginRouting'

class Router extends Component {
  constructor() {
    super()

    this.state = {
      component: null,
      context: null,
      onEnter: null,
      onLeave: null,
      path: null
    }
  }

  componentDidMount() {
    this.props.routes.forEach((route) => {
      const {
        component,
        onEnter,
        onLeave,
        path
      } = route
      createRoute(path, (context) => {
        this.setState({
          component,
          context,
          onEnter,
          onLeave,
          path
        })
      })
    })

    beginRouting()
  }

  /*
   * since this component will update only when the path
   * changes, componentWillUpdate will only be called on
   * route change 
   */
  componentWillUpdate(){
    if(this.state.onLeave){
      this.state.onLeave()
    }
  }
  
  /*
   * since this component will update only when the path
   * changes, componentDidUpdate will only be called on render
   */
  componentDidUpdate(){
    if(this.state.onEnter) {
      this.state.onEnter()
    }
  }

  /*
   * if the current state path matches the nextState path
   * do not update the component
   */
  shouldComponentUpdate(nextProps, nextState){
    return this.state.path !== nextState.path
  }

  render() {
    if(!this.state.component) {
      return null
    }

    return createElement(this.state.component, {
      route: this.state.context
    })
  }
}

Router.propTypes = {
  routes: PropTypes.array.isRequired
}

export default Router
