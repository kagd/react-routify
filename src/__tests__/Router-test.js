jest.mock('../createRoute')
jest.mock('../beginRouting')

import React from 'react'
import renderer from 'react-test-renderer'
import Router from '../Router'
import createRoute from '../createRoute'
import beginRouting from '../beginRouting'

test('default state should render null', () => {
  const Home = () => <div>home</div>
  const routes = [{ path: '/', component: Home }]
  const component = renderer.create(<Router routes={routes} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('beginRouting should be called', () => {
  const Home = () => <div>home</div>
  const routes = [{ path: '/', component: Home }]
  renderer.create(<Router routes={routes} />)
  expect(beginRouting).toBeCalled()
})

test('navigate should update the component', () => {
  const createRouteCallbacks = {}

  createRoute.mockImplementation((path, callback) => {
    createRouteCallbacks[path] = callback
  })

  const Home = () => <div>home</div>
  const About = () => <div>about</div>

  const routes = [
    { path: '/', component: Home },
    { path: '/about', component: About }
  ]

  const component = renderer.create(<Router routes={routes} />)

  let tree

  createRouteCallbacks['/']()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  createRouteCallbacks['/about']()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('route information should exist on props', () => {
  const createRouteCallbacks = {}

  createRoute.mockImplementation((path, callback) => {
    createRouteCallbacks[path] = callback
  })

  const Home = ({ route }) => <div>{route.params.info}</div>

  const routes = [
    { path: '/', component: Home },
  ]

  const component = renderer.create(<Router routes={routes} />)

  createRouteCallbacks['/']({params: {info: 'shinfo'}})
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
