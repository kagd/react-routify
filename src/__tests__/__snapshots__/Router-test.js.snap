exports[`test default state should render null 1`] = `null`;

exports[`test navigate should update the component 1`] = `
<div>
  home
</div>
`;

exports[`test navigate should update the component 2`] = `
<div>
  about
</div>
`;

exports[`test route information should exist on props 1`] = `
<div>
  shinfo
</div>
`;
